### PR TITLE
WIP: client interface

### PR DESCRIPTION
**Pull request by diogomonica**
DO NOT MERGE, THIS IS A WIP.

Attempting to refactor notary cli into libnotary and changing root generation to be independent from repository generation.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+	signer       *signed.Signer
+	tufRepo      *tuf.TufRepo
+	fileStore    store.MetadataStore
+	privKeyStore trustmanager.FileStore
+}
+
+// Target represents a simplified version of the data TUF operates on.
+type Target struct {
+	Name   string
+	Hashes data.Hashes
```
Comment (dmcgowan) [inline client/client.go]: This still exposes a TUF data type on the published interface `data.Hashes`. Ideally we could integrate a docker package such as `distribution/digest`.

Comment (endophage) [inline client/client.go]: A receiver of a Target object shouldn't need to know anything about TUF though right? `data.Hashes` is just a `map[string][]byte` with some convenience functions defined on the `[]byte`

File: client/client.go [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+
+// Update forces TUF to download the remote timestamps and verify if there are
+// any remote changes.
+func (r *NotaryRepository) Update() error {
+	return nil
+}
+
+// Initialize creates a new repository by using rootKey as the root Key for the
+// TUF repository.
+func (r *NotaryRepository) Initialize(rootKey *data.PublicKey) error {
```
Comment (endophage) [inline client/client.go]: If we're aiming to reduce the dependencies of calling code on tuf, this should probably take `rootKeyID string`

File: client/client.go [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+	ListTargets() ([]*Target, error)
+	GetTargetByName(name string) (*Target, error)
+
+	Publish() error
+}
+
+// NewClient is a helper method that returns a new notary Client, given a config
+// file. It makes the assumption that the base directory for the config file will
+// be the place where trust information is being cached locally.
+func NewClient(configFile string) (*NotaryClient, error) {
```
Comment (endophage) [inline client/client.go]: This shouldn't be doing config parsing. Pass in the config values themselves. Both our commands and docker/docker already do config parsing.

File: client/client.go [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+
+// ErrRepositoryNotExist gets returned when trying to make an action over a repository
+/// that doesn't exist
+var ErrRepositoryNotExist = errors.New("repository does not exist")
+
+// Client is the interface that defines the Notary Client type
+type Client interface {
+	ListPrivateKeys() []data.PrivateKey
+	GenRootKey(passphrase string) (*data.PublicKey, error)
+	GetRepository(gun string, baseURL string, transport http.RoundTripper) (Repository, error)
```
Comment (endophage) [inline client/client.go]: Why use a RoundTripper? Does it afford some benefit to us over just doing an `http.Get`?

Comment (diogomonica) [inline client/client.go]: @dmcgowan this one is yours.

File: client/client.go (hunk 3 above)
Comment (diogomonica) [inline client/client.go]: Agreed. I'm tired. Changing it.

File: client/client.go (hunk 2 above)
Comment (diogomonica) [inline client/client.go]: And then create a NewRole with the string directly for root?

File: client/client.go [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+	privKeyStore, err := trustmanager.NewKeyFileStore(filepath.Join(c.baseDir, privDir))
+	if err != nil {
+		return nil, err
+	}
+
+	signer := signed.NewSigner(NewCryptoService(gun, privKeyStore))
+
+	return &NotaryRepository{Gun: gun,
+		baseURL:          baseURL,
+		tufRepoPath:      filepath.Join(c.baseDir, tufDir),
```
Comment (endophage) [inline client/client.go]: `filepath.Join(c.baseDir, tufDir, gun)`

File: client/client.go [hunk 6]
```diff
@@ -0,0 +1,576 @@
...
+	GenRootKey(passphrase string) (*data.PublicKey, error)
+	GetRepository(gun string, baseURL string, transport http.RoundTripper) (Repository, error)
+}
+
+// Repository is the interface that represents a Notary Repository
+type Repository interface {
+	Update() error
+	Initialize(key *data.PublicKey) error
+
+	AddTarget(target *Target) error
```
Comment (dmcgowan) [inline client/client.go]: Target does not need to be mutated nor null, we don't need a pointer here

**Comment by dmcgowan** [comment]
Work is progress is currently being done in another branch

**Pull request closed without merging**